sensor/common/registry/lazy_tls_image_registry.go

```diff
@@ -0,0 +1,176 @@
+package registry
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"sync/atomic"
+
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/registries/types"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/pkg/urlfmt"
+	"github.com/stackrox/rox/pkg/utils"
+)
+
+// lazyTLSCheckRegistry is a wrapper around a registry that performs
+// TLS checks on first Metadata invocation.
+type lazyTLSCheckRegistry struct {
+	// source, dataSource, etc. mirror post initialization fields from
+	// pkg/registries/docker/docker.go. These fields are provided
+	// during construction and are assumed to be valid.
+	source           *storage.ImageIntegration
+	dataSource       *storage.DataSource
+	dockerConfig     *storage.DockerConfig
+	url              string
+	registryHostname string
+
+	// creater creates the underlying registry from source during
+	// initialization.
+	creator types.Creator
+
+	// registry is populated after successful initialization.
+	registry types.Registry
+
+	// tlsCheckCache performs and caches registry TLS checks.
+	tlsCheckCache *tlsCheckCacheImpl
+
+	// initialized tracks whether lazy initialization has completed.
+	initialized      atomic.Uint32
+	initializedMutex sync.RWMutex
+
+	// initError holds the most recent initialization error.
+	initError error
+}
+
+var _ types.ImageRegistry = (*lazyTLSCheckRegistry)(nil)
+
+// Config will NOT trigger initialization, however after successful
+// initialization the values may change.
```

Comment:
Is this potentially dangerous?

Reply:
Not with the current usage, only the `RegistryHostname` field is used and that does not change post initialization (one of the reasons for re-using the `RegistryHostnameURL` method). 

Only the Sensor `registry_store` uses the lazy factory, and currently only in Sensor is there a need to perform the lazy TLS checks, the usages look like this:

```go
...
if r.Config().GetRegistryHostname() == reg {
   return r, nil
}
...
```

If we were to do the lazy init in `Config()` then we lose a large benefit of lazy init - since `Config()`is called on every registry in the store regardless if it ever gets used.